docs/proposals/configmap-rollout.md

```diff
@@ -0,0 +1,181 @@
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# ConfigMap management
+
+## Abstract
+
+A proposal for implementing management of ConfigMaps in the context of a
+[Deployment](http://kubernetes.io/docs/user-guide/deployments/).
+
+Users should be able to have their apps automatically pick up changes to
+their configuration and also have the ability to rollback to previous
+configuration similar to the ability to rollback to previous pod template
+revisions.
+
+Currently, in order to apply updated configuration into a running app, users
+need to 1) either do it manually: copy the ConfigMap that holds their old
+behavior, update it, store it back as a different ConfigMap, and mount the
+new ConfigMap into the Deployment, or 2) edit the original ConfigMap and
+trigger the deployment controller to start a new ReplicaSet (either by
+adding a dummy environment value in the pod template of the Deployment or
+by including the update as part of another update) losing the ability to
+rollback to previous configuration.
+
+This doc proposes a solution for removing the burden of ConfigMap management
+from users by providing a mechanism that will work out of the box without
+having to manually rollout new ReplicaSets.
+
+## Implementation
+
+### Proposed solution
+
+Users should only ever need to create and update a single ConfigMap "foo".
+```sh
+$ kubectl create configmap foo --from-file=config.yml
+configmap "foo" created
+```
+
+We need a way to link "foo" to the Deployment that is going to be triggered
+every time "foo" is updated.
+```sh
+$ kubectl set triggers deployment/bar configmap/foo
+deployment "bar" updated
+```
+
+A `{resource-0}.kubernetes.io/triggered-by: {resource-1}/{name-1}` annotation
+will be added in the resource we want to trigger automatic updates on. In the
+example above, a `deployment.kubernetes.io/triggered-by: configmap/foo`
+annotation should be added in Deployment "bar". We may want to extended this
+mechanism to support cross-namespace triggering since vendors may want to
+trigger on various resource updates not necessarily found in the same namespace
+as the resource that expects to be triggered.
+
+As today, the ConfigMap needs to be mounted as a volume in the Deployment.
+```sh
+$ kubectl set volume deployment/bar -t configmap --configmapname foo
+deployment "bar" updated
+```
+
+A new controller needs to watch for Deployments and ConfigMaps. Whenever it
+spots a Deployment with a `triggered-by` annotation and a mounted ConfigMap,
+it will create a copy of the ConfigMap and mount the copy in the Deployment.
```

Comment:
All the controller has to do is find out which configmap volume it needs to update. The easiest way I can think of is to have `kubectl set trigger` take a container name and store it as a second annotation. Then the controller inspects all configmap volumes and updates the one that points to the container with the given name. The container should already have the volumeMount in place.